linked_list/linked_list.py

```diff
@@ -86,32 +146,82 @@ def visit(self):
 
     # method to reverse the singly linked list
     # note: the nodes should be moved and not just the values in the nodes
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def reverse(self):
-        pass
+        # first node will have no next
+        # second node's next will be the first node
+        # third node's next will be the second node and so on
+
+        # iterative solution:
+        if not self.head:
+            return None # error for empty LL
+        previous = None
+        current = self.head
+        while current:
+            nxt = current.next # this node's original next
+            current.next = previous # this node's original previous
+            previous = current # next node's previous
+            current = nxt # go to saved node
+        self.head = previous
+
   
     ## Advanced/ Exercises
     # returns the value at the middle element in the singly linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def find_middle_value(self):
-        pass
+        list_len = self.length() # O(1)
+        tar_ind = list_len // 2
+        curr_ind = 0
+        current = self.head
+        while curr_ind != tar_ind:
+            curr_ind += 1
+            current = current.next
+        return current.value
 
     # find the nth node from the end and return its value
     # assume indexing starts at 0 while counting to n
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def find_nth_from_end(self, n):
-        pass
+        if self.head is None:
+            return None
+        length = self.length() # O(n)
+        target_position = length - n
+        if target_position < 0:
+            return None
+        position = 1
+        current = self.head
+        while current and position != target_position:
+            current = current.next
+            position += 1
+        if current == None:
+            return None
+        return current.value
 
     # checks if the linked list has a cycle. A cycle exists if any node in the
     # linked list links to a node already visited.
     # returns true if a cycle is found, false otherwise.
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)
     def has_cycle(self):

```

Comment:
🌸 Awesome! Finding cycles in constant space is hard! One method is Floyd's cycle finding algorithm: https://www.geeksforgeeks.org/floyds-cycle-finding-algorithm/ 

You have a slow pointer that traverses the list one node at the time, and another that traverses the list two nodes at a time. If they meet, there's a cycle! It's a tricky solution that I wouldn't expect you to come up with yourself.